test/e2e/util.go

```diff
@@ -214,20 +234,114 @@ func rootCredentials(kubeClient client.Client, name string) (map[string][]byte,
 	return secret.Data, nil
 }
 
-func lookupCNAMEMiekg(host, server string) ([]string, error) {
-	c := dns.Client{}
-	m := dns.Msg{}
-	m.SetQuestion(host+".", dns.TypeCNAME)
+func customResolver(nameserver string) *net.Resolver {
+	return &net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			d := net.Dialer{
+				Timeout: dialTimeout,
+			}
+			if len(nameserver) > 0 {
+				return d.DialContext(ctx, network, fmt.Sprintf("%s:53", nameserver))
+			}
+			return d.DialContext(ctx, network, address)
+		},
+	}
+}
+
+func assertIngressControllerDeleted(t *testing.T, cl client.Client, ing *operatorv1.IngressController) {
+	t.Helper()
+	if err := deleteIngressController(t, cl, ing, 2*time.Minute); err != nil {
+		t.Fatalf("WARNING: cloud resources may have been leaked! failed to delete ingresscontroller %s: %v", ing.Name, err)
+	} else {
+		t.Logf("deleted ingresscontroller %s", ing.Name)
+	}
+}
+
+func operatorConditionMap(conditions ...operatorv1.OperatorCondition) map[string]string {
+	conds := map[string]string{}
+	for _, cond := range conditions {
+		conds[cond.Type] = string(cond.Status)
+	}
+	return conds
+}
+
+func waitForIngressControllerCondition(t *testing.T, cl client.Client, timeout time.Duration, name types.NamespacedName) error {
+	conditions := []operatorv1.OperatorCondition{
+		{Type: "Admitted", Status: operatorv1.ConditionTrue},
+		{Type: operatorv1.IngressControllerAvailableConditionType, Status: operatorv1.ConditionTrue},
+		{Type: operatorv1.LoadBalancerManagedIngressConditionType, Status: operatorv1.ConditionFalse},
+		{Type: operatorv1.DNSManagedIngressConditionType, Status: operatorv1.ConditionFalse},
+	}
+	return wait.PollImmediate(15*time.Second, timeout, func() (bool, error) {
+		ic := &operatorv1.IngressController{}
+		if err := cl.Get(context.TODO(), name, ic); err != nil {
+			t.Logf("failed to get ingresscontroller %s: %v", name.Name, err)
+			return false, nil
+		}
+		expected := operatorConditionMap(conditions...)
+		current := operatorConditionMap(ic.Status.Conditions...)
+		return conditionsMatchExpected(expected, current), nil
+	})
+}
+
+func deleteIngressController(t *testing.T, cl client.Client, ic *operatorv1.IngressController, timeout time.Duration) error {
+	t.Helper()
+	name := types.NamespacedName{Namespace: ic.Namespace, Name: ic.Name}
+	if err := cl.Delete(context.TODO(), ic); err != nil {
+		return fmt.Errorf("failed to delete ingresscontroller: %v", err)
+	}
+
+	err := wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
+		if err := cl.Get(context.TODO(), name, ic); err != nil {
+			if errors.IsNotFound(err) {
+				return true, nil
+			}
+			t.Logf("failed to delete ingress controller %s/%s: %v", ic.Namespace, ic.Name, err)
+			return false, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("timed out waiting for ingresscontroller to be deleted: %v", err)
+	}
```

Comment:
You don't trust Kube API?

Reply:
kube APi will create the instance but in few instance , the conditions were not updated so , test is failed sporadically.
to avoid the sporadically this check added.

Reply:
I don't see the link with the deletion which this method is for. I don't think we need this code, if API server told us the object is deleted we don't have any reason to not trust it.
If we would follow this logic, we have to check the deletion everywhere: when we delete `ExternalDNS` instance, `Route`, `Service`, etc.

Reply:
sure , removed this code and used API alone.